test: add a stress and a soak test for the https outcalls feature

mihailjianu1 opened the pull request:
This test adds an update message to the already existing `proxy_canister`, which performs `count` concurrent outcalls. 

However, as [the number of canister messages is limited to 500](https://internetcomputer.org/docs/building-apps/canister-management/resource-limits), the requests have to be split into batches.

All requests are lightweight both in therms of request size and response size. 

The test sets up three testnets, with 13, 28 and 40 nodes respectively. For each of them it tries to send 200, 500 and 1000 requests, and measures the average qps for each (over 2 experiments).

Results are as follows:
```
QPS:
node# \ concurrency_level | 200 | 500 | 1000
13:                       | 52  |  85 | 81
28:                       | 26  |  53 | 55
40:                       | 51  |  67 | 65
```

The 13 and 28 nodes subnets were set up to better simulate production subnets (package drops, rtts etc). The 28 node subnet saw a ~25% decrease in qps when simulated under production conditions. Meaning we can infer that a ~25% decrease in qps could be seen on the 40 nodes subnet too, in a production environment, leading to ~40-45 qps.

Interpretation:
- The canister measures the time of the http_request to the management canister. This includes ingress message processing, consensus, block making etc etc.
- The 40 node subnet is slower mainly because consensus takes longer, and there's a better chance of some adapters slightly slower
- Not fully saturating the 500 concurrency limit yields worse qps.
- In the current setup, the qps is also correlated with the delay of the target server. 

Colocating nodes doesn't seem to affect the average qps by a lot (sometimes it's ~10% higher / lower, but there's no clear indication that it's way better / worse), as the main bottleneck here is the 500 limit on the concurrent canister messages. 

Future attempts at stressing the feature may include: 

1. Installing multiple proxy canisters on the same subnet to bypass the 500 limit
2. Fully saturating the bottleneck by continuously sending new requests in the message queue instead of waiting for all 500 to return.
3. Sending requests to different servers, so that the warm up is no longer helping much.

Sawchord reviewed (approved):
Awesome! Thanks!

kpop-dfinity commented:
Non-blocking comment: I haven't looked at the code yet but two more things to consider would be:
1) Run the test on the performance cluster to simulate the production environment more closely(see [for example](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/consensus/consensus_performance.rs?L104-123)). I'm not entirely sure if it's going to work because in your setup you need 53+ nodes and I don't know how big the performance cluster is.
2) Impose some artificial network conditions ([example](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/consensus/consensus_performance.rs?L76-81)) to see how the QPS correlates with the latency between nodes. If you want to be even fancier, you might want to use [ProductionSubnetTopology::UZR34](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/driver/src/driver/simulate_network.rs?L172) to simulate a subnet which has some heavy https outcalls traffic.

kpop-dfinity commented on rs/tests/networking/canister_http_stress_test.rs:
```diff
@@ -0,0 +1,366 @@
+    url: String,
+    logger: &Logger,
+    concurrent_requests: u64,
+) -> Result<u64, anyhow::Error> {
```
maybe it would be easier to use if we return `std::time::Duration` here?
```suggestion
) -> Result<Duration, anyhow::Error> {
```

mihailjianu1 replied:
Done, much better. thanks

gbrel commented on rs/tests/networking/canister_http_stress_test.rs:
```diff
@@ -0,0 +1,366 @@
+        total_duration_ns += single_call_duration_ns;
+    }
+
+    let elapsed_seconds = total_duration_ns as f64 / NS_IN_1_SEC as f64;
```
sanity check: Is `total_duration_ns` expected to be identical in value to a hyphotetical `duration` computed as below?
```
    let start = Instant::now();
    while something {
        ...
    }
    let duration = start.elapsed();
```

mihailjianu1 replied:
Not really. Measuring the duration in the test will also include ingress processing etc. Measuring it in the canister gives an accurate duration from the canister's PoV

gbrel commented on rs/tests/networking/canister_http_stress_test.rs:
```diff
@@ -0,0 +1,366 @@
+                    average_latency_s: duration_s,
+                });
+            }
+            leave_proxy_canister_running(&proxy_canister, url.clone(), logger.clone()).await;
```
I am not sure I understand the use of `leave_proxy_canister_running`. Is it a "helper" that will allow us to see performance metrics in prometheus / grafana? 

The way it looks to me is that we have 2 separate types of tests, one through `test_proxy_canister` and another through `leave_proxy_canister_running`. The former serializes the results and the latter doesn't. The latter seems to run continously until a timeout from the proxy canister occurs? 

If my understanding is correct I was wondering whether it makes sense to split these 2 into separate tests. The former being a system test while the second maybe being a dynamic testnet.

mihailjianu1 replied:
You're right' it's better to split. We now have canister_http_stress_test and canister_http_soak_test.

mihailjianu1 commented on rs/tests/networking/canister_http_stress_test.rs:
```diff
@@ -0,0 +1,366 @@
+        .add_subnet(
+            Subnet::new(SubnetType::System)
+                .with_features(SubnetFeatures {
+                    http_requests: true,
```
Every IC needs at least one system subnet to install the NNS.

gbrel replied:
I am referring to `http_requests: true`. Being more specific I also ran the test with:
```
            Subnet::new(SubnetType::System)
                .with_features(SubnetFeatures::default())
```

The above setup still works IIUC because we do not rely on https_outcalls feature on the system subnet in this specific stress test. I think there is value in seeing "http_requests: true" configured only in the relevant subnets.

mihailjianu1 replied:
I see. I removed that now. Funny enough, `SubnetFeauture::default()` actually has `http_requests: true`.

mihailjianu1 commented on rs/tests/networking/canister_http_stress_test.rs:
```diff
@@ -0,0 +1,366 @@
+
+    let app_subnets = get_all_application_subnets(&env);
+
+    for (i, subnet_snapshot) in app_subnets.into_iter().enumerate() {
```
But it is a loop. We go through all application subnets.

mihailjianu1 commented on rs/tests/networking/canister_http_stress_test.rs:
````diff
@@ -0,0 +1,366 @@
+
+Goal:: Measure the qps of http_requests originating from one canister. The test shuold be run with the following command:
+```
+ict ict testnet create canister_http_stress_test --lifetime-mins=180 --output-dir=./canister_http_stress_test -- --test_tmpdir=./canister_http_stress_test
````
yes thanks

mihailjianu1 commented:
> Great stuff!
> 
> Non-blocking comment: I haven't looked at the code yet but two more things to consider would be:
> 
> 1. Run the test on the performance cluster to simulate the production environment more closely(see [for example](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/consensus/consensus_performance.rs?L104-123)). I'm not entirely sure if it's going to work because in your setup you need 53+ nodes and I don't know how big the performance cluster is.
> 2. Impose some artificial network conditions ([example](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/consensus/consensus_performance.rs?L76-81)) to see how the QPS correlates with the latency between nodes. If you want to be even fancier, you might want to use [ProductionSubnetTopology::UZR34](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/driver/src/driver/simulate_network.rs?L172) to simulate a subnet which has some heavy https outcalls traffic.

I can't reply directly to that comment.. so here it is.:
1. I added the performance cluster; it seems to have a similar qps, which I guess means the main bottleneck is still the message queue.
2. I added production tolopology settings to a 13 and a 28 nodes subnet. Only the NNS has 40 nodes currently, so I left the test without any settings. I added some guestimates in the description to how it would behave if the settings were similar to the 28 node uzr34.

mihailjianu1 commented on rs/tests/networking/canister_http_stress_test.rs:
```diff
@@ -0,0 +1,366 @@
+
+            // Make an http_outcall once, to establish the session between the adapter and the target server.
+            // This is necessary in order to avoid the server potentially being overloaded by 40 * 500 TCP/TLS handshake requests.
+            test_proxy_canister(&proxy_canister, url.clone(), logger.clone(), 1).await;
```
Yes, we can get @Sawchord's and @kpop-dfinity's opinion's too. 
My opinion is that we should explain the current performance transparently; meaning we explicitly say that a warm up request is necessary if the adapter has not established a session yet with the target server, and after that, we can _guarantee_ a certain QPS. Only specifying a qps that takes into account that handshake could be misleading and counterintuitive (it goes up with the number of total requests, for example.).

kpop-dfinity replied:
Sorry, I missed the message. 

I think in this particular case, where we bombard a single server with requests, it's okay to warm up the adapter. 
Maybe we could introduce a bool constant `WITH_WARM_UP` which would control whether we want to warm up the adapter or not. Just in case someone wants to run the experiment without warming up.

It could be interesting to see the performance when every request targets a different server, so that the adapter would have to establish a new session for each request. (not in this PR though)

mihailjianu1 commented on rs/tests/networking/canister_http/canister_http.rs:
```diff
@@ -227,6 +234,42 @@ pub fn create_proxy_canister_with_name<'a>(
     Canister::new(runtime, CanisterId::unchecked_from_principal(principal_id))
 }
+
+pub fn create_proxy_canister_with_name_and_cycles<'a>(
```
I'd rather rely on the test driver to do the relevant funneling. Our functions are pretty much just wrappers for their helpers, and some additional logging.

gbrel replied:
>I'd rather rely on the test driver to do the relevant funneling
That is a good point and I agree with you on that.

I will then change this thread to a *nit*, asking what do you think about updating with a cycles param the existing function "create_proxy_canister_with_name" as opposed to creating another "create_proxy_canister_with_name_and_cycles". Pasting a codepilot suggestion here:
```
pub fn create_proxy_canister_with_name<'a>(
    env: &TestEnv,
    runtime: &'a Runtime,
    node: &IcNodeSnapshot,
    canister_name: &str,
    cycles: Option<Cycles>,
) -> Canister<'a> {
    match cycles {
        Some(c) => info!(
            &env.logger(),
            "Installing proxy_canister with a custom cycle amount ({c:?})."
        ),
        None => info!(&env.logger(), "Installing proxy_canister."),
    }

    // Create proxy canister with specified or maximum cycles
    let rt = tokio::runtime::Runtime::new().expect("Could not create tokio runtime.");
    let wasm = load_wasm(env::var("PROXY_WASM_PATH").expect("PROXY_WASM_PATH not set"));
    
    let proxy_canister_id = if let Some(c) = cycles {
        rt.block_on(create_and_install_with_cycles(
            &node.build_default_agent(),
            node.effective_canister_id(),
            &wasm,
            c,
        ))
    } else {
        rt.block_on(create_and_install(
            &node.build_default_agent(),
            node.effective_canister_id(),
            &wasm,
        ))
    };

    // Log appropriate message based on whether cycles were specified
    match cycles {
        Some(c) => info!(
            &env.logger(),
            "Proxy canister {:?} installed with cycles {:?}", proxy_canister_id, c
        ),
        None => info!(
            &env.logger(),
            "proxy_canister {} installed", proxy_canister_id
        ),
    };

    let principal_id = PrincipalId::from(proxy_canister_id);

    // write proxy canister id to TestEnv
    env.write_json_object(canister_name, &principal_id)
        .expect("Could not write proxy canister id to TestEnv.");

    Canister::new(runtime, CanisterId::unchecked_from_principal(principal_id))
}
```

The reason I am suggesting this is that I feel that this function has enough complexity as it is, an duplicating the logic from it in another function can lead to bugs. Nonetheless it's a nit, that can be followed up at a later point.

mihailjianu1 replied:
hmm particularly to that `create_proxy_canister_with_name`, I think it's better to have two small functions than one large one with a lot of if/match statements. Also I don't think that that function has a lot of complexity. Pretty much all it's doing is forwarding the request to the test driver and some additional logging. 

The reason I'm hesitant to call one from another, for example, is that only using test driver's `create_and_install_with_cycles` in our tests would somehow imply we need to explicitly specify the number of cycles, as our tests require that. But only the stress test actually deals with cycles (tangentially).

kpop-dfinity commented on rs/tests/networking/canister_http_stress_test.rs:
```diff
@@ -0,0 +1,231 @@
+    logger: &Logger,
+    concurrent_requests: u64,
+) -> Result<Duration, anyhow::Error> {
+    let proxy_canister = proxy_canister.clone();
+    let url = url.clone();
```
are these clones necessary?

mihailjianu1 replied:
nope. thanks!

kpop-dfinity commented on rs/tests/networking/canister_http_stress_test.rs:
```diff
@@ -0,0 +1,231 @@
+        experiments,
+        qps
+    );
+    (qps, total_duration.checked_div(experiments as u32).unwrap())
```
since we `unwrap` anyways we might as well do an unchecked div:
```suggestion
    (qps, total_duration / experiments as u32)
```

mihailjianu1 replied:
done

mihailjianu1 commented on rs/tests/networking/canister_http_stress_test.rs:
```diff
@@ -0,0 +1,231 @@
+    concurrent_requests: u64,
+) -> (f64, Duration) {
+    let mut experiments = 0;
+    let max_experiments = 2;
```
That's true. There's a caveat that `total_duration` is not really the total duration of the experiment, but rather the total duration experienced by the canister (in addition, the test driver experiences other delays: ingress message processing etc). So especially for small setups (low concurrency, low subnet size), the "experiment time" can be considerably larger than the `total_time`. But I guess it shouldn't be an issue after all. I increased the time out per test from the default 10 minutes to 20 minutes.

mihailjianu1 commented on rs/tests/networking/canister_http_stress_test.rs:
```diff
@@ -0,0 +1,231 @@
+/* tag::catalog[]
+Title:: Stress test for the http_requests feature
+
+Goal:: Measure the qps of http_requests originating from one canister. The test shuold be run with the following command:
```
done

mihailjianu1 commented on rs/tests/networking/canister_http_soak_test.rs:
```diff
@@ -0,0 +1,134 @@
+
+    let app_subnets = get_all_application_subnets(&env);
+
+    for subnet_snapshot in app_subnets.into_iter() {
```
done

mihailjianu1 commented on rs/tests/networking/canister_http_soak_test.rs:
```diff
@@ -0,0 +1,134 @@
+/* tag::catalog[]
+Title:: Soak test for the http_requests feature
+
+Goal:: Measure the evolving qps of http_requests originating from one canister. The test shuold be run with the following command:
```
done
